Universal/Sniffs/CodeAnalysis/MixedBooleanOperatorSniff.php

```diff
@@ -0,0 +1,107 @@
+<?php
+/**
+ * PHPCSExtra, a collection of sniffs and standards for use with PHP_CodeSniffer.
+ *
+ * @package   PHPCSExtra
+ * @copyright 2021 WoltLab GmbH, 2023 PHPCSExtra Contributors
+ * @license   https://opensource.org/licenses/LGPL-3.0 LGPL3
+ * @link      https://github.com/PHPCSStandards/PHPCSExtra
+ */
+
+namespace PHPCSExtra\Universal\Sniffs\CodeAnalysis;
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+
+/**
+ * Forbid mixing `&&` and `||` within a single expression without making precedence
+ * clear using parentheses.
+ *
+ * @link https://github.com/squizlabs/PHP_CodeSniffer/pull/3205
+ * @link https://github.com/php-fig/per-coding-style/issues/22
+ *
+ * @since 1.2.0
+ */
+final class MixedBooleanOperatorSniff implements Sniff
+{
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @since 1.2.0
+     *
+     * @return array<int|string>
+     */
+    public function register()
+    {
+        return [
+            \T_BOOLEAN_OR,
+            \T_BOOLEAN_AND,
+        ];
+    }
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @since 1.2.0
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file being scanned.
+     * @param int                         $stackPtr  The position of the current token
+     *                                               in the stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $tokens = $phpcsFile->getTokens();
+        $token  = $tokens[$stackPtr];
+
+        $start = $phpcsFile->findStartOfStatement($stackPtr);
+
+        if ($token['code'] === T_BOOLEAN_AND) {
+            $search = T_BOOLEAN_OR;
+        } elseif ($token['code'] === T_BOOLEAN_OR) {
+            $search = T_BOOLEAN_AND;
+        } else {
+            throw new \LogicException('Unreachable');
+        }
+
+        while (true) {
+            $previous = $phpcsFile->findPrevious(
+                [
+                    $search,
+                    T_OPEN_PARENTHESIS,
+                    T_OPEN_SQUARE_BRACKET,
+                    T_CLOSE_PARENTHESIS,
+                    T_CLOSE_SQUARE_BRACKET,
```

Comment:
Would the use of `PHPCSUtils\Tokens\Collections::shortArrayTokens()` make it harder to upstream the Sniff? After all, this is my main goal. 

> Some example code (will probably need better and more test cases, but this is just to trigger your imagination and yes, I know it's silly code, but people write silly code....):

Thanks, added this as a test case.

> Also: if we can come up with any conceivable code sample where an attribute would be in an expression containing booleans and/or operators, it would probably be a good idea to also skip over attributes.

Thanks, I hate writing such code, but there is indeed a code sample that allows for attributes. It's already correctly handled.

Reply:
> Would the use of `PHPCSUtils\Tokens\Collections::shortArrayTokens()` make it harder to upstream the Sniff? After all, this is my main goal.

Not really, it's just shorthand for the tokens used for short arrays. For upstream, you'd need to explicitly add those tokens, but PHPCSExtra can use the shorthand.

https://github.com/PHPCSStandards/PHPCSUtils/blob/77a6aa4e6abc437eb0cb71702c1a635655a33af8/PHPCSUtils/Tokens/Collections.php#L499-L502

> Thanks, I hate writing such code, but there is indeed a code sample that allows for attributes. It's already correctly handled.

Yup... I've learned to write really, really, really horrible code just to test sniffs, the joys of maintaining tooling like this... 😂